manifests/init.pp

```diff
@@ -164,10 +175,33 @@
     restart    => '/usr/bin/systemctl reload nftables',
   }
 
+  if $allow_unmanaged_rules {
+    file { $inmem_rules_hash_file:
+      ensure => absent,
+    }
+  } else {
+    exec { 'Reload nftables if there are un-managed rules':
+      command     => '/usr/bin/systemctl reload nftables',
+      refreshonly => false,
+      unless      => "/usr/bin/test -s ${inmem_rules_hash_file} -a \"$(nft -s list ruleset | sha1sum)\" = \"$(cat ${inmem_rules_hash_file})\"",
```

Comment:
The `nft -s list ruleset` will also contain any chains that are present in the `nftables::noflush_chains` parameter.

Taking the typical use case of fail2ban maintaining a chain we expect that chain to drift almost constantly so ideally we want to ignore those chains when making the comparison to avoid unnecessary reloads.

The reloads themselves will be fine since these noflush chains are already ignored by the reload.

Reply:
A hard one. The easy exit is to make them mutually exclusive, otherwise we'd have to find a way to remove the dynamic rules when Puppet takes a snapshot of what's in memory. However, having an option that purges unmanaged rules allowing unmanaged rules sounds odd to me so I'm in favour of the mutual exclusivity.

Reply:
If we added --json to the dump filtering out the noflush chains to the dump should be doable?

Reply:
I would certainly use both options if they were not mutually exclusive.

Reply:
> If we added --json to the dump filtering out the noflush chains to the dump should be doable?

It's doable, however this would introduce a dependency on a JSON processor, like `jq`. Unfortunately the objects are not nested when dumped with `--json`, so the script would have to iterate over `nftables::noflush_tables` and for each element attach a filter like:

```
| del(.[] | select(.rule.table == "table_name" or .chain.table == "table_name" or .table.name == "table_name"))
```
The filter would probably have to filter by table name **and** family, as it's legal to have two tables with the same name and different families, making it even uglier.

Ugly but it should work. The script to hash the rules could be used for both dumping (to be called by Systemd as currently proposed) and comparing (to be called by the `Exec`).

Another option is to do something similar to [puppet.nft.epp](https://github.com/voxpupuli/puppet-nftables/blob/master/templates/config/puppet.nft.epp#L7) and generate an array of managed tables by removing `nftables::noflush_tables` from `$facts['nftables']['tables']`. Then sort the array, hash each table independently and then join the hashes and use that string to detect changes. Something like (pseudobash):

```bash
tables=("inet filter" "ip nat" "ip6 nat") # To be generated by Puppet
hashes=""

for table in "${tables[@]}"
do
  sha1=$(/sbin/nft -s list table $table | /usr/bin/sha1sum | cut -f 1 -d " ")
  hashes+=$sha1
done

if [ $2 == "store" ];
then
  /bin/echo $hashes > $1
elif [ $2 == "compare" ];
then
  /usr/bin/test -s $1 -a "$(cat $1)" = "$hashes"
else
  echo "Dunno what to do"
  exit 1
fi
```
We'd have to think here if the boostrapping makes sense as `$facts['nftables']['tables']` won't be available during the first run.